### ISSUE TITLE
EVP_PKEY refcount leak

### ISSUE THREAD

**Issue by dwmw2** [OpenSC/libp11#111]
Since commit 70eaa8855 we have a rampant `EVP_PKEY` leak. Testing with something like:
```
openssl pkeyutl -engine pkcs11 -inkey 'pkcs11:token=openconnect-test;id=%01;pin-value=1234' \
     -sign -in foo -out bar  -keyform engine
```

Every call to `pkcs11_rsa()` will increase the refcount on the `EVP_PKEY`.

**Comment by dwmw2** [comment]
This helps...

```
--- a/src/p11_rsa.c
+++ b/src/p11_rsa.c
@@ -37,9 +37,12 @@ static int rsa_ex_index = 0;
 static RSA *pkcs11_rsa(PKCS11_KEY *key)
 {
        EVP_PKEY *evp_key = pkcs11_get_key(key, key->isPrivate);
+       RSA *rsa;
        if (evp_key == NULL)
                return NULL;
-       return EVP_PKEY_get0_RSA(evp_key);
+       rsa = EVP_PKEY_get0_RSA(evp_key);
+       EVP_PKEY_free(evp_key);
+       return rsa;
 }
 
 /* PKCS#1 v1.5 RSA signature */
```

But I'm confused about the refcounting and object lifetime rules here. We store a reference to the `EVP_PKEY` in the `PKCS11_KEY`... so now my test process exits with one refcount still on the `EVP_PKEY`. When is the `PKCS11_KEY` supposed to be freed?

The usage model after #110 is supposed to be that a single `EVP_PKEY` can outlive all other references to the engine, and in `pkcs11_rsa_free_method()` when we're done with the key, we drop the last refcount on the engine.

But that never gets called because we have a recursive refcount; the `PKCS11_KEY` is only ever freed when the engine is cleaned up... and the engine is never cleaned up because the `pkcs11_rsa_free_method()` is never called... and that never happens because the `PKCS11_KEY` never got freed.

**Comment by dwmw2** [comment]
Do I read this correctly that we only ever destroy the `PKCS11_KEY` from `pkcs11_destroy_keys()` where we do it for a whole token at a time?

**Comment by mtrojnar** [comment]
You are right: `pkcs11_rsa()` needs to free its EVP_PKEY reference.

Before https://github.com/OpenSC/libp11/commit/70eaa8855824e149133943c52a2c4aa47ccb5272 it only didn't crash and burn because `pkeyutl.c` never calls `ENGINE_finish()`...  In other words we had a way more severe bug that was hidden by missing engine cleanup.

> Do I read this correctly that we only ever destroy the PKCS11_KEY from pkcs11_destroy_keys() where we do it for a whole token at a time? 

Yes, we only destroy our PKCS11_KEY structures at `PKCS11_login()`, `PKCS11_logout()`, and `PKCS11_release_all_slots()`.  These operations invalidate any PKCS11_KEY pointers returned by `PKCS11_enumerate_keys()`.

**Comment by dwmw2** [comment]
OK, so now we're just left with the recursive refcount between the `EVP_PKEY` *(via the final `ENGINE_finish()`)* and the `PKCS11_KEY`.

One way we might fix that is by having the `EVP_PKEY` pin a *structural* reference on the engine, not a *functional* reference. That means that our `engine_finish()` method will happen while the `EVP_PKEY` still exists, but *not* `engine_destroy()`.

So all the slots and tokens and (unused) cache `PKCS11_KEY` and certs and stuff can all be torn down in `engine_finish()` as they are at the moment. All we need to do is ensure that the `PKCS11_KEY`, session and other bits we need for the remaining `EVP_PKEY` is not freed until we're done with it — we need to fix up the refcounts to work in that direction.

**Comment by mtrojnar** [comment]
> That means that our `engine_finish()` method will happen while the `EVP_PKEY` still exists, but not `engine_destroy()`.

Then we need to move the whole body of `pkcs11_finish()` to `pkcs11_destroy()` (in `eng_back.c`), because `EVP_PKEY` depends on the associated `PKCS11_KEY`, which in turn is destroyed by `PKCS11_release_all_slots()`.

Guessing which `PKCS11_KEY` are "used" and which are not is a very dangerous idea.  In fact, when we start destroying PKCS11_KEY objects, their pointers returned by `PKCS11_enumerate_keys()` become invalid.  This would break any code that uses `PKCS11_enumerate_keys()` (which is pretty much any code that uses libp11).

I guess we should open a separate issue for this discussion.

**Comment by dwmw2** [comment]
> Then we need to move the whole body of `pkcs11_finish()` to `pkcs11_destroy()` (in `eng_back.c`), because `EVP_PKEY` depends on the associated `PKCS11_KEY`, which in turn is destroyed by `PKCS11_release_all_slots()`.

Well, doing that cleanup in `pkcs11_finish()` was actually the whole point of the exercise, because cleaning up the `PKCS11_KEY` is currently the only way the `EVP_PKEY` gets released... and the final refcount on the engine can't go away until the final `EVP_PKEY` is released.

(But see below for a different idea which might be simpler)

> Guessing which PKCS11_KEY are "used" and which are not is a very dangerous idea. In fact, when we start destroying PKCS11_KEY objects, their pointers returned by PKCS11_enumerate_keys() become invalid. This would break any code that uses PKCS11_enumerate_keys() (which is pretty much any code that uses libp11).

Right. I'd covered that with the fairly understated words *"need to fix up the refcounts to work in that direction"* :)

And yes, it would be.... nontrivial. Although *perhaps* we want to sort all that out anyway, if we want to submit code to OpenSSL upstream for 1.$NEXT, and have it pass review? 

Another thought... what about a model where in `engine_finish()` we only iterate over all the `PKCS11_KEY` objects and remove *our* refcount to their `EVP_PKEY`?

So any `EVP_PKEY` which is just cached internally will die at that point, and any associated refcount to the engine will go with it. So in the common case, the engine's structural refcount will also reach zero right then, and it'll be unloaded. But any `EVP_PKEY` which is "out there" in the application and has an additional refcount, will continue to pin the engine in place until it's freed.

Now, there are some details we have to worry about here too. Since we use it for things like `pkcs11_get_key_size()` we probably need to keep the `EVP_PKEY *` in the `PKCS11_KEY` even after `engine_finish()` when we don't have our *own* refcount to it, and that means we probably want to augment our key's finish method *(e.g. `pkcs11_rsa_free_method()`)* also to remove the `EVP_PKEY *` from the corresponding `PKCS11_KEY` because now we know the `EVP_PKEY` really *is* going away and we have a stale pointer. But *maybe* we can make that model work with a little less effort?

> I guess we should open a separate issue for this discussion.

Perhaps. The *specific* issue above is basically resolved. Or since there are a lot of inter-related things to think about, any *specific* issue isn't quite the right place; maybe we should just take it to email? Or improved latency, how about IRC? We have the #opensc channel on Freenode, right? I was never in there, but I  am now.
